File: backend/src/main/java/org/netra/features/events/entity/DonationEvent.java
package org.netra.features.events.entity;

import jakarta.persistence.*;
import java.time.Instant;
import java.util.UUID;

@Entity
@Table(name = "donation_events")
public class DonationEvent {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "blood_bank_id", nullable = false)
    private UUID bloodBankId;

    @Column(name = "title", nullable = false, length = 255)
    private String title;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(name = "event_type", nullable = false, length = 64)
    private DonationEventType eventType = DonationEventType.BLOOD_DONATION_CAMP;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 32)
    private DonationEventStatus status = DonationEventStatus.DRAFT;

    @Column(name = "venue_name", nullable = false, length = 255)
    private String venueName;

    @Column(name = "address", nullable = false, length = 255)
    private String address;

    @Column(name = "city", nullable = false, length = 100)
    private String city;

    @Column(name = "state", nullable = false, length = 100)
    private String state;

    @Column(name = "postal_code", nullable = false, length = 20)
    private String postalCode;

    @Column(name = "latitude", nullable = false)
    private Double latitude;

    @Column(name = "longitude", nullable = false)
    private Double longitude;

    @Column(name = "start_at", nullable = false)
    private Instant startAt;

    @Column(name = "end_at", nullable = false)
    private Instant endAt;

    @Column(name = "registration_open_at", nullable = false)
    private Instant registrationOpenAt;

    @Column(name = "registration_close_at", nullable = false)
    private Instant registrationCloseAt;

    @Column(name = "donor_capacity", nullable = false)
    private Integer donorCapacity;

    @Column(name = "current_registration_count", nullable = false)
    private Integer currentRegistrationCount = 0;

    @Column(name = "created_by", nullable = false)
    private UUID createdBy;

    @Column(name = "created_at", nullable = false)
    private Instant createdAt = Instant.now();

    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt = Instant.now();

    @Column(name = "published_at")
    private Instant publishedAt;

    @Column(name = "cancelled_at")
    private Instant cancelledAt;

    @Column(name = "cancelled_by")
    private UUID cancelledBy;

    @Version
    @Column(name = "version", nullable = false)
    private Long version = 0L;

    public DonationEvent() {
    }

    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        this.id = id;
    }

    public UUID getBloodBankId() {
        return bloodBankId;
    }

    public void setBloodBankId(UUID bloodBankId) {
        this.bloodBankId = bloodBankId;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public DonationEventType getEventType() {
        return eventType;
    }

    public void setEventType(DonationEventType eventType) {
        this.eventType = eventType;
    }

    public DonationEventStatus getStatus() {
        return status;
    }

    public void setStatus(DonationEventStatus status) {
        this.status = status;
    }

    public String getVenueName() {
        return venueName;
    }

    public void setVenueName(String venueName) {
        this.venueName = venueName;
    }

    public String getAddress() {
        return address;
    }

    public void setAddress(String address) {
        this.address = address;
    }

    public String getCity() {
        return city;
    }

    public void setCity(String city) {
        this.city = city;
    }

    public String getState() {
        return state;
    }

    public void setState(String state) {
        this.state = state;
    }

    public String getPostalCode() {
        return postalCode;
    }

    public void setPostalCode(String postalCode) {
        this.postalCode = postalCode;
    }

    public Double getLatitude() {
        return latitude;
    }

    public void setLatitude(Double latitude) {
        this.latitude = latitude;
    }

    public Double getLongitude() {
        return longitude;
    }

    public void setLongitude(Double longitude) {
        this.longitude = longitude;
    }

    public Instant getStartAt() {
        return startAt;
    }

    public void setStartAt(Instant startAt) {
        this.startAt = startAt;
    }

    public Instant getEndAt() {
        return endAt;
    }

    public void setEndAt(Instant endAt) {
        this.endAt = endAt;
    }

    public Instant getRegistrationOpenAt() {
        return registrationOpenAt;
    }

    public void setRegistrationOpenAt(Instant registrationOpenAt) {
        this.registrationOpenAt = registrationOpenAt;
    }

    public Instant getRegistrationCloseAt() {
        return registrationCloseAt;
    }

    public void setRegistrationCloseAt(Instant registrationCloseAt) {
        this.registrationCloseAt = registrationCloseAt;
    }

    public Integer getDonorCapacity() {
        return donorCapacity;
    }

    public void setDonorCapacity(Integer donorCapacity) {
        this.donorCapacity = donorCapacity;
    }

    public Integer getCurrentRegistrationCount() {
        return currentRegistrationCount;
    }

    public void setCurrentRegistrationCount(Integer currentRegistrationCount) {
        this.currentRegistrationCount = currentRegistrationCount;
    }

    public UUID getCreatedBy() {
        return createdBy;
    }

    public void setCreatedBy(UUID createdBy) {
        this.createdBy = createdBy;
    }

    public Instant getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(Instant createdAt) {
        this.createdAt = createdAt;
    }

    public Instant getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(Instant updatedAt) {
        this.updatedAt = updatedAt;
    }

    public Instant getPublishedAt() {
        return publishedAt;
    }

    public void setPublishedAt(Instant publishedAt) {
        this.publishedAt = publishedAt;
    }

    public Instant getCancelledAt() {
        return cancelledAt;
    }

    public void setCancelledAt(Instant cancelledAt) {
        this.cancelledAt = cancelledAt;
    }

    public UUID getCancelledBy() {
        return cancelledBy;
    }

    public void setCancelledBy(UUID cancelledBy) {
        this.cancelledBy = cancelledBy;
    }

    public Long getVersion() {
        return version;
    }

    public void setVersion(Long version) {
        this.version = version;
    }
}

File: frontend/lib/features/events/models/donation_event.dart
enum DonationEventType {
  bloodDonationCamp;

  static DonationEventType fromString(String? value) {
    switch (value?.toUpperCase()) {
      case 'BLOOD_DONATION_CAMP':
      default:
        return DonationEventType.bloodDonationCamp;
    }
  }

  String get displayName {
    switch (this) {
      case DonationEventType.bloodDonationCamp:
        return 'Blood Donation Camp';
    }
  }
}

enum DonationEventStatus {
  draft,
  pendingApproval,
  published,
  registrationClosed,
  ongoing,
  completed,
  cancelled,
  rejected;

  static DonationEventStatus fromString(String? value) {
    switch (value?.toUpperCase()) {
      case 'PENDING_APPROVAL':
        return DonationEventStatus.pendingApproval;
      case 'PUBLISHED':
        return DonationEventStatus.published;
      case 'REGISTRATION_CLOSED':
        return DonationEventStatus.registrationClosed;
      case 'ONGOING':
        return DonationEventStatus.ongoing;
      case 'COMPLETED':
        return DonationEventStatus.completed;
      case 'CANCELLED':
        return DonationEventStatus.cancelled;
      case 'REJECTED':
        return DonationEventStatus.rejected;
      case 'DRAFT':
      default:
        return DonationEventStatus.draft;
    }
  }

  String get displayName {
    switch (this) {
      case DonationEventStatus.draft:
        return 'Draft';
      case DonationEventStatus.pendingApproval:
        return 'Pending Approval';
      case DonationEventStatus.published:
        return 'Registration Open';
      case DonationEventStatus.registrationClosed:
        return 'Registration Closed';
      case DonationEventStatus.ongoing:
        return 'Camp Ongoing';
      case DonationEventStatus.completed:
        return 'Completed';
      case DonationEventStatus.cancelled:
        return 'Cancelled';
      case DonationEventStatus.rejected:
        return 'Rejected';
    }
  }
}

class DonationEventSummary {
  final String id;
  final String bloodBankId;
  final String bloodBankName;
  final String title;
  final String? description;
  final DonationEventType eventType;
  final DonationEventStatus status;
  final String venueName;
  final String address;
  final String city;
  final String state;
  final String postalCode;
  final DateTime startAt;
  final DateTime endAt;
  final DateTime registrationOpenAt;
  final DateTime registrationCloseAt;
  final int donorCapacity;
  final int currentRegistrationCount;
  final int remainingCapacity;
  final double? distanceKm;
  final bool isRegistrationOpen;

  const DonationEventSummary({
    required this.id,
    required this.bloodBankId,
    required this.bloodBankName,
    required this.title,
    this.description,
    required this.eventType,
    required this.status,
    required this.venueName,
    required this.address,
    required this.city,
    required this.state,
    required this.postalCode,
    required this.startAt,
    required this.endAt,
    required this.registrationOpenAt,
    required this.registrationCloseAt,
    required this.donorCapacity,
    required this.currentRegistrationCount,
    required this.remainingCapacity,
    this.distanceKm,
    required this.isRegistrationOpen,
  });

  factory DonationEventSummary.fromJson(Map<String, dynamic> json) {
    return DonationEventSummary(
      id: json['id'] as String,
      bloodBankId: json['bloodBankId'] as String,
      bloodBankName: json['bloodBankName'] as String? ?? 'NETRA Blood Centre',
      title: json['title'] as String,
      description: json['description'] as String?,
      eventType: DonationEventType.fromString(json['eventType'] as String?),
      status: DonationEventStatus.fromString(json['status'] as String?),
      venueName: json['venueName'] as String,
      address: json['address'] as String,
      city: json['city'] as String,
      state: json['state'] as String,
      postalCode: json['postalCode'] as String,
      startAt: DateTime.parse(json['startAt'] as String),
      endAt: DateTime.parse(json['endAt'] as String),
      registrationOpenAt: DateTime.parse(json['registrationOpenAt'] as String),
      registrationCloseAt:
          DateTime.parse(json['registrationCloseAt'] as String),
      donorCapacity: json['donorCapacity'] as int,
      currentRegistrationCount: json['currentRegistrationCount'] as int? ?? 0,
      remainingCapacity: json['remainingCapacity'] as int? ?? 0,
      distanceKm: json['distanceKm'] != null
          ? (json['distanceKm'] as num).toDouble()
          : null,
      isRegistrationOpen: json['isRegistrationOpen'] as bool? ?? false,
    );
  }
}

class DonationEventDetail extends DonationEventSummary {
  final double? latitude;
  final double? longitude;
  final DateTime? publishedAt;
  final DateTime? cancelledAt;
  final String? cancellationReason;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  const DonationEventDetail({
    required super.id,
    required super.bloodBankId,
    required super.bloodBankName,
    required super.title,
    super.description,
    required super.eventType,
    required super.status,
    required super.venueName,
    required super.address,
    required super.city,
    required super.state,
    required super.postalCode,
    required super.startAt,
    required super.endAt,
    required super.registrationOpenAt,
    required super.registrationCloseAt,
    required super.donorCapacity,
    required super.currentRegistrationCount,
    required super.remainingCapacity,
    super.distanceKm,
    required super.isRegistrationOpen,
    this.latitude,
    this.longitude,
    this.publishedAt,
    this.cancelledAt,
    this.cancellationReason,
    this.createdAt,
    this.updatedAt,
  });

  factory DonationEventDetail.fromJson(Map<String, dynamic> json) {
    return DonationEventDetail(
      id: json['id'] as String,
      bloodBankId: json['bloodBankId'] as String,
      bloodBankName: json['bloodBankName'] as String? ?? 'NETRA Blood Centre',
      title: json['title'] as String,
      description: json['description'] as String?,
      eventType: DonationEventType.fromString(json['eventType'] as String?),
      status: DonationEventStatus.fromString(json['status'] as String?),
      venueName: json['venueName'] as String,
      address: json['address'] as String,
      city: json['city'] as String,
      state: json['state'] as String,
      postalCode: json['postalCode'] as String,
      startAt: DateTime.parse(json['startAt'] as String),
      endAt: DateTime.parse(json['endAt'] as String),
      registrationOpenAt: DateTime.parse(json['registrationOpenAt'] as String),
      registrationCloseAt:
          DateTime.parse(json['registrationCloseAt'] as String),
      donorCapacity: json['donorCapacity'] as int,
      currentRegistrationCount: json['currentRegistrationCount'] as int? ?? 0,
      remainingCapacity: json['remainingCapacity'] as int? ?? 0,
      distanceKm: json['distanceKm'] != null
          ? (json['distanceKm'] as num).toDouble()
          : null,
      isRegistrationOpen: json['isRegistrationOpen'] as bool? ?? false,
      latitude: json['latitude'] != null
          ? (json['latitude'] as num).toDouble()
          : null,
      longitude: json['longitude'] != null
          ? (json['longitude'] as num).toDouble()
          : null,
      publishedAt: json['publishedAt'] != null
          ? DateTime.parse(json['publishedAt'] as String)
          : null,
      cancelledAt: json['cancelledAt'] != null
          ? DateTime.parse(json['cancelledAt'] as String)
          : null,
      cancellationReason: json['cancellationReason'] as String?,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'] as String)
          : null,
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'] as String)
          : null,
    );
  }
}
